### PR TITLE
Unsim Block Info Functions

### PR DESCRIPTION
**Pull request by kantai**
This PR removes the simming code for block-info and replaces it with actual information from the header database. This PR touches a couple of different spots:

1. Clarity instances need to be passed a handle to the headers_db
2. `at-block` _cannot_ accept the stacks header hash as an input -- this is non-unique and is a value that has no meaning to the MARF. This PR adds an `id-hash-header` option to `get-block-info` to allow smart contracts to use that value. An alternative is to use the block-height, and keep the `id-hash-header` hidden. If this is preferred, I can change that.
3. The MARF needs to be able to figure out what block height the current block is, even if it's not the "miner_tip", the clearest example is for evaluating something like `(at-block <foo> block-height)`, the block that the MARF _switches_ to is not the miner_tip, but it still needs to return a correct value for `block-height`. In order to do that, this PR just sets a static key to the current block height for every MARF block (this replaces the code that attempts to detect when the miner_tip is the correct result).
4. In testing the integration of all these pieces, the testnet `run_loop` callback signature was updated to allow the tests to run clarity code snippets when new chainstate arrives, and passes in the block header of the new block.

This PR _doesn't_ yet implement fetching the miner address or block time -- fetching those would require either providing the Clarity VM instance with a handle to the BurnDB in addition to the headers db, or shoving that information into the headers db, and I didn't want to go down either of those paths without getting @jcnelson's thoughts on that first.

### CODE REVIEW

File: src/chainstate/stacks/index/marf.rs [hunk 1]
```diff
@@ -607,6 +600,9 @@ impl MARF {
         test_debug!("Set {}::{} = {}", BLOCK_HEIGHT_TO_HASH_MAPPING_KEY, height, next_block_hash.to_hex());
         test_debug!("Set {}::{} = {}", BLOCK_HASH_TO_HEIGHT_MAPPING_KEY, next_block_hash.to_hex(), height);
 
+        keys.push(OWN_BLOCK_HEIGHT_KEY.to_string());
```
Comment (jcnelson) [inline src/chainstate/stacks/index/marf.rs]: Can you do me a favor and add a `test_debug!()` like those above to show that we're inserting `OWN_BLOCK_HEIGHT_KEY` here?

File: src/testnet/run_loop.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@ pub struct RunLoop {
     nodes: Vec<Node>,
     new_burnchain_state_callback: Option<fn(u8, &BurnchainState)>,
     new_tenure_callback: Option<fn(u8, &LeaderTenure)>,
-    new_chain_state_callback: Option<fn(u8, &StacksChainState)>,
+    new_chain_state_callback: Option<fn(u8, &mut StacksChainState, &BlockHeaderHash)>,
```
Comment (jcnelson) [inline src/testnet/run_loop.rs]: Not saying that we have to do this in this particular PR, but I'm increasingly uncomfortable with these being a set of callbacks that may or may not be `Some(...)`, since it introduces the need for a lot of runtime boiilerplate checks.  Can either you or @lgalabru open an issue to make `RunLoop` a `trait` instead, and then refactor this code into one or more `impl`s of that trait with the requisite behaviors?  I can see there being at least two `impl`s for example -- one for a private "regtest" and one for a public "testnet".

File: src/util/db.rs [hunk 3]
```diff
@@ -155,6 +155,13 @@ macro_rules! impl_byte_array_from_column {
                 Ok(inst)
             }
         }
+
+        impl rusqlite::types::ToSql for $thing {
```
Comment (jcnelson) [inline src/util/db.rs]: Ah, if we're going to do this, then we should probably go and remove all the calls to `$thing.to_hex() as &dyn ToSql` in the database code.

File: src/vm/database/clarity_db.rs [hunk 4]
```diff
@@ -39,32 +41,89 @@ pub enum StoreType {
...
+            .map(|x| x.anchored_header.block_hash())
+    }
+    
+    fn get_burn_header_hash_for_block(&self, id_bhh: &BlockHeaderHash) -> Option<BurnchainHeaderHash> {
+        get_stacks_header_info(self, id_bhh)
+            .map(|x| x.burn_header_hash)
+    }
+
+    fn get_burn_block_time_for_block(&self, _id_bhh: &BlockHeaderHash) -> Option<u64> {
+        panic!("Block time data not available in burn header db")
```
Comment (jcnelson) [inline src/vm/database/clarity_db.rs]: Adding this is going to take more than adding a column for it in the headers db.  In particular, the code in `src/burnchains` will need to be updated so that the timestamp is something the burnchain indexer picks up from burnchain block headers.  We might want to just leave that for a separate PR.

Comment (jcnelson) [reply src/vm/database/clarity_db.rs]: Alternatively, I can contribute to this PR and add it.  Open to either approach.

File: src/vm/database/clarity_db.rs [hunk 5]
```diff
@@ -143,84 +202,42 @@ impl <'a> ClarityDatabase <'a> {
...
-        VRFSeed::from_bytes(&self.get_simmed_block(block_height).vrf_seed).unwrap()
+    pub fn get_current_block_height(&mut self) -> u32 {
+        self.store.get_current_block_height()
     }
 
-    fn set_simmed_block_height(&mut self, block_height: u64) {
-        let block_height_key = ClarityDatabase::make_key_for_trip(
-            &QualifiedContractIdentifier::transient(), StoreType::SimmedBlockHeight, ":dummy:");
-        self.put(&block_height_key, &block_height);
+    pub fn get_block_header_hash(&mut self, block_height: u32) -> BlockHeaderHash {
```
Comment (jcnelson) [inline src/vm/database/clarity_db.rs]: What happens when a contract tries to get the block header hash for a block height that's too high?

Comment (jcnelson) [reply src/vm/database/clarity_db.rs]: Same question for all these un-simmed methods.

File: src/vm/docs/mod.rs [hunk 6]
```diff
@@ -581,10 +581,13 @@ const AT_BLOCK: SpecialAPI = SpecialAPI {
     description: "The `at-block` function evaluates the expression `expr` _as if_ it were evaluated at the end of the
 block indicated by the _block-hash_ argument. The `expr` closure must be read-only.
 
+Note: The block identifying hash must be a hash returned by the `id-header-hash` block information
+property. This hash uniquely identifies Stacks blocks, whereas `header-hash` is not necessarily unique.
```
Comment (jcnelson) [inline src/vm/docs/mod.rs]: Might want to clarify that `id-header-hash` is globally unique across all Stacks forks, since it's the hash of both the Stacks block and the burnchain block that selected it.  The `header-hash` value is unique in the context of a single fork.

File: src/vm/tests/integrations.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    contract_addr: &StacksAddress, contract_name: &str,
+    function_name: &str, function_args: &[Value]) -> Vec<u8> {
+
+    let contract_name = ContractName::from(contract_name);
+    let function_name = ClarityName::from(function_name);
+
+    let payload = TransactionContractCall {
+        address: contract_addr.clone(),
+        contract_name, function_name,
+        function_args: function_args.iter().map(|x| x.clone()).collect()
```
Comment (jcnelson) [inline src/vm/tests/integrations.rs]: Not addressed in this PR, but is there a maximum allowed length for the number of arguments to a Clarity function?  Or is it bound only by the compute budget?

File: src/vm/types/mod.rs [hunk 8]
```diff
@@ -24,7 +24,7 @@ pub struct TupleData {
     pub data_map: BTreeMap<ClarityName, Value>
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
```
Comment (jcnelson) [inline src/vm/types/mod.rs]: Why no `Debug`?

File: src/util/db.rs (hunk 3 above)
Comment (kantai) [reply src/util/db.rs]: Okay, will do

File: src/vm/database/clarity_db.rs (hunk 4 above)
Comment (kantai) [reply src/vm/database/clarity_db.rs]: I'll just leave it as a panicking stub function for now -- I'll add a stub for the miner address as well (#1131)

File: src/vm/database/clarity_db.rs (hunk 5 above)
Comment (kantai) [reply src/vm/database/clarity_db.rs]: If it gets to this function, it panics. The caller of these functions performs a check on the block_height and returns None if the block_height is >= the current block height.

File: src/vm/tests/integrations.rs (hunk 7 above)
Comment (kantai) [reply src/vm/tests/integrations.rs]: Currently only bound by the compute budget.

File: src/vm/docs/mod.rs (hunk 6 above)
Comment (kantai) [reply src/vm/docs/mod.rs]: Sure

File: src/chainstate/stacks/index/marf.rs (hunk 1 above)
Comment (kantai) [reply src/chainstate/stacks/index/marf.rs]: Sure

File: src/testnet/run_loop.rs (hunk 2 above)
Comment (kantai) [reply src/testnet/run_loop.rs]: I don't generally think its a good idea to design a trait before there are actually multiple implementations of the trait. Otherwise, we end up thinking we know what the common interface should be before actually knowing what it should be.

For the boiler plate checks on these callbacks, I can just refactor these checks into function calls that handle the Option matching.

File: src/util/db.rs (hunk 3 above)
Comment (kantai) [reply src/util/db.rs]: Updated in 7a9fb081c1ed9d76003793c3b39c2e08bc14c3aa -- judging by the vast majority of the remaining `.to_hex()` usage, do you think we should just `impl Fmt::Display` for these types as well ?

File: src/testnet/run_loop.rs (hunk 2 above)
Comment (jcnelson) [reply src/testnet/run_loop.rs]: Sounds good.  I'm fine with deferring making this a trait until we have two non-trivial impls, but I'd nevertheless appreciate moving the Option matching to one place for the time being.

File: src/util/db.rs (hunk 3 above)
Comment (jcnelson) [reply src/util/db.rs]: Thanks for doing that!  I'm fine with doing this for `fmt::Display` as well, if you're up for it.

File: src/vm/database/clarity_db.rs (hunk 5 above)
Comment (jcnelson) [reply src/vm/database/clarity_db.rs]: Alright, just making sure we're not setting ourselves up for unexpected runtime panics :+1:

File: src/vm/tests/integrations.rs (hunk 7 above)
Comment (jcnelson) [reply src/vm/tests/integrations.rs]: Thanks!  I'll make a note of that in the transaction serialization and processing code as well.

File: src/util/db.rs (hunk 3 above)
Comment (kantai) [reply src/util/db.rs]: Yep, will add Fmt::Display for this as well.

File: src/chainstate/stacks/index/marf.rs (hunk 1 above)
Comment (kantai) [reply src/chainstate/stacks/index/marf.rs]: Addressed in f2063f7ef72f3501806c53c3b6f2b12c43a00b93

File: src/testnet/run_loop.rs (hunk 2 above)
Comment (kantai) [reply src/testnet/run_loop.rs]: Addressed in d43618a12c87932389a5e41ec3a575bc6e342f8d

File: src/util/db.rs (hunk 3 above)
Comment (kantai) [reply src/util/db.rs]: Updated in 566f1be5d341032eddd9309bb41c317af2d9d0f0

File: src/testnet/run_loop.rs (hunk 2 above)
Comment (lgalabru) [inline src/testnet/run_loop.rs]: Why do we have to make the arg `StacksChainState` mutable?

File: src/vm/docs/mod.rs [hunk 9]
```diff
@@ -581,10 +581,13 @@ const AT_BLOCK: SpecialAPI = SpecialAPI {
...
+
 The function returns the result of evaluating `expr`.
 ",
     example: "(at-block 0x0000000000000000000000000000000000000000000000000000000000000000 (var-get data))
-(at-block (get-block-info? header-hash (- block-height u10)) (var-get data))"
+(at-block (get-block-info? id-header-hash (- block-height u10)) (var-get data))"
```
Comment (lgalabru) [inline src/vm/docs/mod.rs]: Not a strong opinion, but I think that passing a block-height to `(at-block ...)` would make this feature more accessible to smart contract developers (+ more terse).

**Comment by kantai** [comment]
Ooops -- looks like I was addressing @jcnelson's comments in a different branch. I cherry-picked the commits over to this one.

File: src/testnet/run_loop.rs (hunk 2 above)
Comment (kantai) [reply src/testnet/run_loop.rs]: To query the chain state information -- since that information is in the MARF, reads require a mutable reference.
